server/handlers/pattern_collab_handler.go

```diff
@@ -0,0 +1,255 @@
+package handlers
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+var upgrader = websocket.Upgrader{}
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 60 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+)
+
+var (
+	newline = []byte{'\n'}
+)
+
+type Client struct {
+	hub       *Hub
+	conn      *websocket.Conn
+	connected bool
+	patternID string
+	send      chan []byte
+}
+
+type Hub struct {
+	patterns                map[string]map[*Client]bool
+	broadcast               chan *SocketMessage
+	register                chan *Client
+	unregister              chan *Client
+	updateSubscribedPattern chan *PatternUpdateRequest
+}
+
+type SocketMessage struct {
+	Type   string      `json:"type"`
+	Topics []string    `json:"topics,omitempty"`
+	Topic  string      `json:"topic,omitempty"`
+	Data   interface{} `json:"data,omitempty"`
+}
+
+type PatternUpdateRequest struct {
+	oldPatternID string
+	newPatternID string
+	client       *Client
+}
+
+func (c *Client) processMessage(message []byte) {
+	sockMessage := &SocketMessage{}
+	err := json.Unmarshal(message, sockMessage)
+	if err != nil {
+		return
+	}
+	switch sockMessage.Type {
+	case "ping":
+		pongMessage := &SocketMessage{Type: "pong"}
+		msg, err := json.Marshal(pongMessage)
+		if err != nil {
+			return
+		}
+		c.send <- msg
+	case "subscribe":
+		newPatternID := sockMessage.Topics[0]
+		updateRequest := &PatternUpdateRequest{
+			oldPatternID: c.patternID,
+			newPatternID: newPatternID,
+			client:       c,
+		}
+		c.patternID = newPatternID
+		c.hub.updateSubscribedPattern <- updateRequest
+	case "unsubscribe":
+		oldPatternID := sockMessage.Topics[0]
+		updateRequest := &PatternUpdateRequest{
+			oldPatternID: oldPatternID,
+			newPatternID: "",
+			client:       c,
+		}
+		c.patternID = ""
+		c.hub.updateSubscribedPattern <- updateRequest
+	case "publish":
+		c.hub.broadcast <- sockMessage
+	}
+}
+
+func (c *Client) readMessages() {
+	_ = c.conn.SetReadDeadline(time.Now().Add(pongWait))
+	c.conn.SetPongHandler(func(string) error { _ = c.conn.SetReadDeadline(time.Now().Add(pongWait)); return nil })
+	defer func() {
+		c.hub.unregister <- c
+		c.conn.Close()
+	}()
+	for {
+		_, message, err := c.conn.ReadMessage()
+		if err != nil {
+			break
+		}
+
+		c.processMessage(message)
+	}
+}
+
+func (c *Client) writeMessages() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		c.hub.unregister <- c
```

Comment:
This same client was send to unregister channel when readMessages() returns. Also the connection is closed when the readMessages() returns. There will be a send() on closed channel. Do both these things at one place.

Reply:
done!